overlord/snapstate/handlers.go

```diff
@@ -777,6 +785,165 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
+			}
+
+			logger.Debugf("Download task %s will wait 1min for pre-download task %s", task.ID(), preTask.ID())
+			return &state.Retry{After: 2 * time.Minute}
+		}
+	}
+
+	return nil
+}
+
+func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	snapsup, theStore, user, err := downloadSnapParams(st, t)
+	if err != nil {
+		return err
+	}
+
+	targetFn := snapsup.MountFile()
+	dlOpts := &store.DownloadOptions{
+		// pre-downloads are only triggered in auto-refreshes
+		IsAutoRefresh: true,
+		RateLimit:     autoRefreshRateLimited(st),
+	}
+
+	perfTimings := state.TimingsForTask(t)
+	st.Unlock()
+	timings.Run(perfTimings, "pre-download", fmt.Sprintf("pre-download snap %q", snapsup.SnapName()), func(timings.Measurer) {
+		err = theStore.Download(tomb.Context(nil), snapsup.SnapName(), targetFn, snapsup.DownloadInfo, nil, user, dlOpts)
+	})
+	st.Lock()
+	if err != nil {
+		return err
+	}
+	perfTimings.Save(st)
+
+	var waitingTasks []string
+	if err := t.Get("waiting-tasks", &waitingTasks); err != nil && !errors.Is(err, &state.NoStateError{}) {
+		return err
+	}
+
+	// there are download tasks waiting for this one so unblock them and we don't
+	// need to spawn a new change
+	if len(waitingTasks) > 0 {
+		for _, taskID := range waitingTasks {
+			st.Task(taskID).At(time.Time{})
+		}
+
+		st.EnsureBefore(0)
+		return nil
+	}
+
+	_, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	info, err := snapst.CurrentInfo()
+	if err != nil {
+		return err
+	}
+
+	// TODO: in the future, do a hard check before starting an auto-refresh so there's
+	// no change of the snap starting between changes and preventing it from going through
+	err = backend.WithSnapLock(info, func() error {
+		return refreshAppsCheck(info)
+	})
+	if err != nil {
+		if !errors.Is(err, &BusySnapError{}) {
+			return err
+		}
+
+		var refreshInfo *userclient.PendingSnapRefreshInfo
+		if err := t.Get("refresh-info", &refreshInfo); err != nil {
+			return err
+		}
+
+		return asyncRefreshOnSnapClose(m.state, refreshInfo)
+	}
+
+	// signal to the autorefresh manager to continue the previously blocked autorefresh
+	st.Cache("continue-auto-refresh", true)
+	st.EnsureBefore(0)
+
+	return nil
+}
+
+// asyncRefreshOnSnapClose asynchronously waits for the snap the close, notifies
+// the user and then triggers an auto-refresh.
+func asyncRefreshOnSnapClose(st *state.State, refreshInfo *userclient.PendingSnapRefreshInfo) error {
+	var monitoredSnaps map[string]bool
+	if cachedMonitored := st.Cached("monitored-snaps"); cachedMonitored != nil {
+		monitoredSnaps, _ = cachedMonitored.(map[string]bool)
+	} else {
+		monitoredSnaps = make(map[string]bool)
+	}
+
+	// there's already a goroutine waiting for this snap to close so just notify
+	if monitoredSnaps[refreshInfo.InstanceName] {
+		asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
```

Comment:
We notify a few lines below too, I wonder if we could always notify here and then return if the monitor is already up? 
```go
--- a/overlord/snapstate/handlers.go
+++ b/overlord/snapstate/handlers.go
@@ -906,9 +906,10 @@ func asyncRefreshOnSnapClose(st *state.State, refreshInfo *userclient.PendingSna
                monitoredSnaps = make(map[string]bool)
        }
 
-       // there's already a goroutine waiting for this snap to close so just notify
+       // notify the user about the blocked refresh
+       asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
+       // there's already a goroutine waiting for this snap
        if monitoredSnaps[refreshInfo.InstanceName] {
-               asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
                return nil
        }
 
@@ -918,9 +919,6 @@ func asyncRefreshOnSnapClose(st *state.State, refreshInfo *userclient.PendingSna
                return fmt.Errorf("cannot monitor for snap closure: %w", err)
        }
 
-       // notify the user about the blocked refresh
-       asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
-
        monitoredSnaps[refreshInfo.InstanceName] = true
        st.Cache("monitored-snaps", monitoredSnaps)
 
```
the difference is of course that it would generate a notification even if cgroupMonitorSnapEnded() has failed so maybe not a good idea :)

Reply:
Yes, the issue is that this is a user-facing action so we run the risk of confusing users if there was an error in the cgroups monitoring call (because we'd notify but nothing would happen afterwards)